### PR TITLE
Tqdm tk

### PR DESCRIPTION
**Pull request by richardsheridan**
This PR adds a Tkinter based GUI for tqdm. It does not attempt to provide the graphs of the current `tqdm_gui`, but it is more performant, and suitable for use on the command line, within fully-fledged Tkinter apps, or even crazier cases, like the [trio guest mode async await tkinter app](https://github.com/richardsheridan/magic-afm/blob/048dbb143edd40c80c65e8e4fa8a5a23e3805f00/magic_gui.py#L236-L372) that I originally built it for. One outstanding bug is that Tkinter does not play well with the `Tmonitor` functionality, so I disabled it by default for now by setting the class attribute `monitor_interval=0`.

This PR does not touch the broader, non-experimental functionality of tqdm.

I would boldly suggest making this class the default `tqdm.gui.tqdm`, as it is lighter-weight and has fewer dependencies compared to the `matplotlib.pyplot`-based implementation. That could be renamed to `tqdm_mpl` or something like that, depending on how committed to backwards compatibility you are for these experimental bits.

### CODE REVIEW

**Comment by richardsheridan** [comment]
Hmmm you think you could overlook that weird __init__ call or should I really pull out a base class for both of them?

File: tqdm/gui.py [hunk 1]
```diff
@@ -316,6 +316,245 @@ def tgrange(*args, **kwargs):
...
+
+        warn('GUI is experimental/alpha', TqdmExperimentalWarning, stacklevel=2
+             )
+        self.tk_dispatching = self.tk_dispatching_helper()
+        if not self.tk_dispatching:
+            # leave is problematic if the mainloop is not running
+            self.leave = False
+
+        self.tk_window.protocol("WM_DELETE_WINDOW", self.cancel)
+        self.tk_window.wm_title("tqdm_tk")
```
Comment (casperdcl) [inline tqdm/gui.py:391]: makes more sense to me to use

```suggestion
        self.tk_window.wm_title(self.desc)
```

and remove `self.tk_desc_var` etc...

Comment (richardsheridan) [reply tqdm/gui.py:391]: I've been using multi-line strings for `desc`, which doesn't work in the title bar. also serious users can reach in and set `wm_title` as well if they want!  

I guess the real question for you is: are multi-line description strings generally supported in `tqdm`? If yes, we have to stick with the `tk_desc_label` stuff.

**Comment by richardsheridan** [comment]
Thanks @casperdcl! I am happy with it right now, so please review the new commits. also, if you have any ideas/suggestions on aesthetic improvements, I'm willing to try to accommodate.

**Comment by casperdcl** [comment]
improvements will probably best be done iteratively; I'm sure planet Earth will chime in.

Can't guarantee when I'll have time to review (I'm not funded for tqdm :P)

**Review by casperdcl — CHANGES_REQUESTED** [review]
Thanks a lot @richardsheridan :confetti_ball: - I've finally reviewed this and also rebased onto the current release.

- split back into original `tqdm.gui` and new `tqdm.tk`
  + added you as main author
  + fixed conflicts, updated inheritance-related things & style fixes
- left 3 `# WIP` comments (see https://github.com/tqdm/tqdm/pull/1006/files)

**Comment by richardsheridan** [comment]
Regarding the WIP comments, I think the right way to do those color changes is with `ttk.Style`, which might lead nicely into changing "states" rather than colours, but I will have to look into it. 

I subclassed refresh to make sure we never interact with the lock without changing the interface, as it is unnecessary for the display. I'm happy to drop it if you think that that is better!

**Comment by casperdcl** [comment]
> I think the right way to do those color changes is with `ttk.Style`, which might lead nicely into changing "states" rather than colours, but I will have to look into it.

yes, that would be great.

> I subclassed refresh to make sure we never interact with the lock without changing the interface, as it is unnecessary for the display. I'm happy to drop it if you think that that is better!

Probably better to not bother overriding. If the tk updates take a long time to process, the default logic (which supports e.g. locks with timout) may well speed things up.

**Comment by richardsheridan** [comment]
Changing the progress bar color requires using a [non-default](https://stackoverflow.com/questions/13510882/how-to-change-ttk-progressbar-color-in-python) theme that may make the bar look less native on some platforms. Also, did you want the text to change color too?

**Comment by casperdcl** [comment]
ideally only the bar colour, not text (as with `tqdm.notebook` and the standard terminal `tqdm.tqdm`/`tqdm.cli`).

However if it's too difficult, completely ignoring the `colour` option is also fine.

**Comment by richardsheridan** [comment]
Using the non-default theme has the knock-on effect of changing the style of an entire tk app if the progress bar is embedded in a larger program. It would be the equivalent of setting an entire jupyter notebook to dark mode just to change the colorbar color, so I think skipping the color customization makes sense.

A very interested user can manipulate the `_tk_pbar` object directly

File: tqdm/tk.py [hunk 2]
```diff
@@ -145,7 +141,11 @@ def display(self):
         self._tk_n_var.set(self.n)
         d = self.format_dict
         d['ncols'] = None
-        self._tk_text_var.set(self.format_meter(**d))
+        text = self.format_meter(**d)
+        # fixup only default bar format
+        if self.bar_format == "{l_bar}{r_bar}":
+            text = text.replace("||", "", 1)
+        self._tk_text_var.set(text)
```
Comment (casperdcl) [inline tqdm/tk.py:148]:
```suggestion
        self._tk_text_var.set(
            text.replace("||", "", 1) if "{l_bar}{r_bar}" in self.bar_format else text)
```

**Review by casperdcl — APPROVED** [review]
great!

**Pull request merged**